smart-contracts/contracts/CosmosBridge.sol

```diff
@@ -179,35 +188,36 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
         require(sigs.length == claims.length, "INV_CLM_LEN");
         require(sigs.length == signatureData.length, "INV_SIG_LEN");
 
+        uint256 intermediateNonce = lastNonceSubmitted;
+        require(intermediateNonce + 1 == claims[0].nonce, "INV_ORD");
+
         for (uint256 i = 0; i < sigs.length; i++) {
+            intermediateNonce++;
             _submitProphecyClaimAggregatedSigs(sigs[i], claims[i], signatureData[i]);
         }
+
+        lastNonceSubmitted = intermediateNonce;
     }
 
     function submitProphecyClaimAggregatedSigs(
         bytes32 hashDigest,
         ClaimData calldata claimData,
         SignatureData[] calldata signatureData
     ) external {
+        uint256 previousNonce = lastNonceSubmitted;
+        require(previousNonce + 1 == claimData.nonce, "INV_ORD");
+
         _submitProphecyClaimAggregatedSigs(hashDigest, claimData, signatureData);
+
+        // update the nonce
+        lastNonceSubmitted = claimData.nonce;
     }
 
-    // Essentially, each tx to the submitProphecyClaimAggregatedSigs needs to include a
-    // networkDescriptor or other unique identifier that only the smart contract on that chain has.
-    // If the chain identifier in the smart contract and the chain identifier in the message do not
-    // match, the transaction should fail.
     function _submitProphecyClaimAggregatedSigs(
         bytes32 hashDigest,
-        ClaimData calldata claimData,
+        ClaimData memory claimData,
```

Comment:
This should be calldata to conserve gas. If it is memory, then it will be copied from calldata to memory which will cost more.

Reply:
I know, but I tried a lot of different approaches to solve the `stack too deep` error and this was the only one that worked.
Local scopes wouldn't help here, as you mentioned in our meeting. Passing the struct instead of separate values didn't work either. I even tried to move some verifications to different places, to no avail.
After the gas optimization, we're still spending less than we were before.